compiler/rustc_hir/src/definitions.rs

```diff
@@ -361,8 +368,20 @@ impl Definitions {
         let disambiguator = {
             let next_disamb = self.next_disambiguator.entry((parent, data)).or_insert(0);
             let disambiguator = *next_disamb;
-            *next_disamb = next_disamb.checked_add(1).expect("disambiguator overflow");
-            disambiguator
+            // The disambiguator number space is split into the ones counting from 0 upwards,
+            match query_local_disambiguator {
+                None => {
+                    assert!(disambiguator < u32::MAX / 2);
+                    *next_disamb = next_disamb.checked_add(1).expect("disambiguator overflow");
+                    disambiguator
+                }
+                // and the ones counting from MAX downwards.
+                Some(local) => {
+                    // Ensure these two number spaces do not collide. 2^31 disambiguators should be enough for everyone.
+                    assert!(local < u32::MAX / 2);
+                    u32::MAX - local
```

Comment:
This affects symbol names in incremental compilation. Not sure if that is a problem